sale_advance_payment/tests/test_sale_advance_payment.py

```diff
@@ -218,56 +214,147 @@ def test_sale_advance_payment(self):
+            "active_ids": [self.sale_order_1.id],
+            "active_id": self.sale_order_1.id,
+        }
+        # Create Advance Payment 2 - USD - cash
         advance_payment_2 = (
             self.env["account.voucher.wizard"]
             .with_context(context_payment)
             .create(
                 {
                     "journal_id": self.journal_usd_cash.id,
-                    "payment_type": "outbound",
-                    "amount_advance": 200,
+                    "amount_advance": 2000,
                     "order_id": self.sale_order_1.id,
                 }
             )
         )
         advance_payment_2.make_advance_payment()
-        self.assertEqual(self.sale_order_1.amount_residual, 3600)
+        pre_payment = self.sale_order_1.account_payment_ids
+        self.assertEqual(len(pre_payment), 1)
+        self.assertEqual(self.sale_order_1.amount_residual, 1600)
+        # generate a partial bill, reconcile with pre payment, check amount residual.
+        self.sale_order_1.action_confirm()
+        self.assertEqual(self.sale_order_1.invoice_status, "to invoice")
+        # Adjust billing method to then do a partial bill with a total amount
+        # smaller than the pre-payment.
+        self.product_1.invoice_policy = "delivery"
+        self.order_line_1.qty_delivered = 10.0
+        self.assertEqual(self.order_line_1.qty_to_invoice, 10.0)
+        self.product_2.invoice_policy = "delivery"
+        self.order_line_2.qty_delivered = 0.0
+        self.assertEqual(self.order_line_2.qty_to_invoice, 0.0)
+        self.product_3.invoice_policy = "delivery"
+        self.order_line_3.qty_delivered = 0.0
+        self.assertEqual(self.order_line_3.qty_to_invoice, 0.0)
+        self.sale_order_1._create_invoices()
+        self.assertEqual(self.sale_order_1.invoice_status, "no")
+        self.assertEqual(self.sale_order_1.amount_residual, 1600)
+        invoice = self.sale_order_1.invoice_ids
+        invoice.invoice_date = fields.Date.today()
+        invoice.action_post()
+        self.assertEqual(invoice.amount_residual, 1200)
+        active_ids = invoice.ids
+        self.env["account.payment.register"].with_context(
+            active_model="account.move", active_ids=active_ids
+        ).create(
+            {
+                "amount": 300.0,
+                "group_payment": True,
+                "payment_difference_handling": "open",
+            }
+        )._create_payments()
+        self.assertEqual(invoice.amount_residual, 900)
+        self.assertEqual(self.sale_order_1.amount_residual, 1300)
+
+        # Partially reconciling the pre-payment should not affect amount_residual in PO.
+        (
+            liquidity_lines,
+            counterpart_lines,
+            writeoff_lines,
+        ) = pre_payment._seek_for_lines()
+        (
+            counterpart_lines
+            + invoice.line_ids.filtered(
+                lambda line: line.account_internal_type == "payable"
```

Comment:
receivable.

you are probably not reconciling anything here, can you add an assert for the amount_residual on the invoice after this?

`self.assertEqual(invoice.amount_residual, 0)`